src/ess/reduce/workflow.py

```diff
@@ -91,3 +93,139 @@ def assign_parameter_values(pipeline: Pipeline, values: dict[Key, Any]) -> Pipel
         else:
             pipeline[key] = value
     return pipeline
+
+
+def prune_nexus_domain_types(
+    workflow: Pipeline,
+    *,
+    targets: Iterable[Key] | None = None,
+    targets_per_run: Iterable[type[Scope]] | None = None,
+    targets_per_run_and_monitor: Iterable[type[ScopeTwoParams]] | None = None,
+    run_types: Iterable[Key] | None = None,
+    monitor_types: Iterable[Key] | None = None,
+) -> Pipeline:
+    """Remove unused types from a workflow.
+
+    This function removes all nodes from a workflow that are not needed to compute
+    given targets with given run and monitor types.
+
+    Warning
+    -------
+    This modifies the input workflow.
+
+    Parameters
+    ----------
+    workflow:
+        Workflow to remove types from.
+    targets:
+        Unparametrized types to keep.
+    targets_per_run:
+        Types parametrized by run type to keep.
+    targets_per_run_and_monitor:
+        Types parametrized by run type and monitor type to keep.
+    run_types:
+        List of run types to include in the workflow. If not provided, all run types
+        are included.
+    monitor_types:
+        List of monitor types to include in the workflow. If not provided, all monitor
+        types are included.
```

Comment:
Before looking at this more closely, can we have a chat about the new interface? I feel the need to specify targets means that this function (and pruning in general) will be unusable by anyone not intricately familiar with a workflow. 

Furthermore, doesn't it mean that every time a workflow is changed there is a risk that all the pruning calls break (and need to be updated)? It may even become hard to see through, e.g., if someone adds a new target (towards the end of a workflow) but the workflow it refuses to compute it because it was unintentionally pruned.